### DISCUSSION TITLE
Support Shared `debug.keystore` for Multi-Workstation Android Development

### DISCUSSION THREAD

**Discussion by vzool** [beeware/briefcase#1944] [Q&A] [answered]
Hi BeeWare team,

I'm working on a [BeeWare project](https://github.com/vzool/zakat-tracker) that involves building across multiple Workstations. I've run into an issue where the app needs to be uninstalled from a mobile before installing a new build from another Workstation. This is due to a mismatch in the debug signing key used for each build.

I'm wondering! if there's a way to configure [briefcase](https://github.com/beeware/briefcase) to use a single, shared `debug.keystore` file across all development machines. This would streamline the testing process significantly by eliminating the need for repeated uninstalls, which can be time-consuming, especially when working with test data.

### Here's my current workflow:

1- Build app on Workstation A.
2- Install on mobile A.
3- Make changes to code.
4- Build app on Workstation B.
5- Uninstall app from mobile A.
6- Install new build on mobile A. (Test data lost)

Ideally, I'd like to avoid step 5 and retain test data across builds from different machines.

### Best I can do

I created the longest validity date for `debug.keystore`
```bash
keytool -genkeypair -alias androiddebugkey -keyalg RSA -keysize 4096 -validity 2739931 -keystore publish/android/debug.keystore -storepass "zakat-tracker" -keypass "zakat-tracker" -dname "CN=Abdelaziz Elrashed Elshaikh Mohamed, OU=IT, O=Zakat Movement, L=Omdurman, S=Khartoum, C=SD"
```
and copied it into `~/.android/debug.keystore` which I guess is the default keystore location
```bash
cp publish/android/debug.keystore ~/.android/debug.keystore 
```
But, it fails
[briefcase.2024_08_08-06_17_40.build.log](https://github.com/user-attachments/files/16537200/briefcase.2024_08_08-06_17_40.build.log)

### Questions:

- Is it possible to configure a shared `debug.keystore` for BeeWare/Toga Android projects?
- If not currently supported, are there any potential workarounds or planned features to address this?
- Could this be achieved by customizing the briefcase build process or configuration?

Any guidance on this would be greatly appreciated.

Thanks!

**Comment by freakboy3742** [comment]
Does such an option exist at present? No. Briefcase doesn't currently have any built-in support for signing Android apps.

As for workarounds - you might be able to use the `ANDROID_USER_HOME` environment variable to redirect the default keystore to a shared location, but I can't offer any "ready to use" advice beyond that.

*Could* this be achieved? Absolutely. Signing is already implemented for macOS and Windows apps; we'd very much like to add it for Android. #1268 is tracking this broader feature; I imagine part of that work would be ensuring that an isolated keystore is used, or that a custom keystore can be specified for signing purposes.

**Reply by mhsmith** [reply]
I've added a workaround to #1268 which uses `build_gradle_extra_content`.

**Reply by vzool** [reply, marked as answer]
It worked! Thanks @mhsmith & @freakboy3742  
Here is the full code at `pyproject.toml` for the community:
```toml
build_gradle_extra_content = """\
android {
    signingConfigs {
        debug {
            keyAlias "android-debug-key"
            keyPassword "zakat-tracker"
            storePassword "zakat-tracker"
            storeFile file("../../../../../publish/android/debug.keystore")
        }
    }
}
"""
```
Thanks

**Reply by mhsmith** [reply]
Based on https://stackoverflow.com/questions/18328730, it looks like you also need to add a `buildTypes` section to tell it to use the config.
